src/ray/common/ray_syncer/ray_syncer-inl.h

```diff
@@ -94,37 +99,178 @@ class NodeSyncConnection {
+/// It keeps track of the message received and sent between two nodes and uses that to
+/// deduplicate the messages. It also supports the batching for performance purposes.
+template <typename T>
+class RaySyncerBidiReactorBase : public RaySyncerBidiReactor, public T {
+ public:
+  /// Constructor of RaySyncerBidiReactor.
+  ///
+  /// \param io_context The io context for the callback.
+  /// \param remote_node_id The node id connects to.
+  /// \param message_processor The callback for the message received.
+  /// \param cleanup_cb When the connection terminates, it'll be called to cleanup
+  ///     the environment.
+  RaySyncerBidiReactorBase(
+      instrumented_io_context &io_context,
+      const std::string &remote_node_id,
+      std::function<void(std::shared_ptr<const RaySyncMessage>)> message_processor)
+      : RaySyncerBidiReactor(remote_node_id),
+        io_context_(io_context),
+        message_processor_(std::move(message_processor)) {}
+
+  bool PushToSendingQueue(std::shared_ptr<const RaySyncMessage> message) override {
+    // Try to filter out the messages the target node already has.
+    // Usually it'll be the case when the message is generated from the
+    // target node or it's sent from the target node.
+    if (message->node_id() == GetRemoteNodeID()) {
+      // Skip the message when it's about the node of this connection.
+      return false;
+    }
+
+    auto &node_versions = GetNodeComponentVersions(message->node_id());
+    if (node_versions[message->message_type()] < message->version()) {
+      node_versions[message->message_type()] = message->version();
+      sending_buffer_[std::make_pair(message->node_id(), message->message_type())] =
+          std::move(message);
+      StartSend();
+      return true;
+    }
+    return false;
+  }
+
+  virtual ~RaySyncerBidiReactorBase() {}
+
+ protected:
+  void StartPull() {
+    receiving_message_ = std::make_shared<RaySyncMessage>();
+    RAY_LOG(DEBUG) << "Start reading: " << NodeID::FromBinary(GetRemoteNodeID());
+    StartRead(receiving_message_.get());
+  }
+
+  /// The io context
+  instrumented_io_context &io_context_;
+
+ private:
   /// Handle the udpates sent from the remote node.
   ///
   /// \param messages The message received.
-  void ReceiveUpdate(RaySyncMessages messages);
+  void ReceiveUpdate(std::shared_ptr<const RaySyncMessage> message) {
+    auto &node_versions = GetNodeComponentVersions(message->node_id());
+    RAY_LOG(DEBUG) << "Receive update: "
+                   << " message_type=" << message->message_type()
+                   << ", message_version=" << message->version()
+                   << ", local_message_version="
+                   << node_versions[message->message_type()];
+    if (node_versions[message->message_type()] < message->version()) {
+      node_versions[message->message_type()] = message->version();
+      message_processor_(message);
+    }
+  }
+
+  void SendNext() {
+    sending_ = false;
+    StartSend();
```

Comment:
Because it'll only send when there is data. Pull will wait until there is data but send you need data to send. gRPC will wait until there is data to be sent.